File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.cpp
 class Solution {
public:
    int singleNonDuplicate(vector<int>& nums) {
        int l=0,r=nums.size()-1,ans=-1; 
        while(l<=r){
            int mid = l+(r-l)/2;
            cout<<mid<<endl;
            int left_val=mid-1<0?INT_MIN:nums[mid-1];
            int right_val=mid+1==nums.size()?INT_MAX:nums[mid+1];
            if(nums[mid]!=left_val && nums[mid]!=right_val) return nums[mid];
            if(nums[mid]==right_val){
                int size=r-(mid+1);
                if(size%2==0) r=mid-1;
                else l=mid+2;
            }
            else{
                int size = mid-1;
                if(size%2) r=mid-2;
                else l=mid+1;
            }
        }
        return -1;
    }
};

File: 0540-single-element-in-a-sorted-array/0540-single-element-in-a-sorted-array.py
class Solution:
    def singleNonDuplicate(self, nums: List[int]) -> int:
        # get the mid 
        # if the compare with left and right if its not equal thats the resul 
        # if prev is same exclude that and get the cnt 
        l,r = 0,len(nums)-1 
        while l<=r: 
            mid=l+(r-l)//2 
            prevele = nums[mid-1] if mid-1>=0 else float("-inf")
            nextele = nums[mid+1] if mid+1<len(nums) else float("-inf")

            if nums[mid]!=prevele and nums[mid]!=nextele: 
                return nums[mid]
            
            leftcnt = mid-l
            leftcnt = leftcnt-1 if nums[mid]==prevele else leftcnt
            leftcnt = max(0,leftcnt)
            
            if leftcnt%2!=0:
                r = mid-2 if nums[mid]==prevele else mid -1 
            else: 
                l= mid+2 if nums[mid]==nextele else  mid+1 
            
        return -1 